### PR TITLE
Update ESLint config and dependencies

### PR DESCRIPTION
**Pull request by pjanik**
A few additions to ESLint config:
```javascript
  "jsx-quotes": ["error", "prefer-double"],
  "object-curly-spacing": ["error", "always"],
  "array-bracket-spacing": ["error", "never"],
```
- `jsx-quotes` to be conistent with the main `quotes` rule
- `object-curly-spacing` and `array-bracket-spacing` are set so they match the .eslntrc.js file itself. :wink: I have no strong opinion about a particular config, but it's nice to have these rules, as we can call linter with `--fix` argument, and fix all the files automatically.

More important update is that I've added linting of the cypress files and used cypress and jest eslint plugins. Config is at the bottom and it overrides the main one:

```javascript
    { // rules specific to Jest tests
      files: ["**/*.test.*"],
      env: {
        node: true,
        jest: true
      },
      plugins: ["jest"],
      extends: ["plugin:jest/recommended"],
      rules: {
        "@typescript-eslint/no-require-imports": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "jest/no-done-callback": "off"
      }
    },
    { // rules specific to Cypress tests
      files: ["cypress/**"],
      env: {
        node: true,
        "cypress/globals": true
      },
      plugins: ["cypress"],
      extends: ["plugin:cypress/recommended"],
      rules: {
        "@typescript-eslint/no-require-imports": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "cypress/no-unnecessary-waiting": "off"
      }
    },
```

Remembering our tests in real projects, I've disabled `cypress/no-unnecessary-waiting` and `jest/no-done-callback`, but otherwise recommended settings are used.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,93 +1,120 @@
...
+    "prefer-const": "error",
+    "prefer-object-spread": "error",
+    "prefer-regex-literals": "error",
+    "prefer-rest-params": "error",
+    "prefer-spread": "error",
+    "quotes": ["error", "double", { allowTemplateLiterals: true, avoidEscape: true }],
+    "jsx-quotes": ["error", "prefer-double"],
+    "object-curly-spacing": ["error", "always"],
+    "array-bracket-spacing": ["error", "never"],
+    "radix": "error",
```
Comment (kswenson) [inline .eslintrc.js:68]: I believe the pattern has been to alphabetize the rules, so these rules should be distributed accordingly.

File: .eslintrc.js [hunk 2]
```diff
@@ -1,93 +1,120 @@
...
+    "react/jsx-curly-spacing": ["error", { "when": "never", "children": { "when": "always" } }],
```
Comment (kswenson) [inline .eslintrc.js:69]:
```suggestion
    "react/jsx-curly-spacing": ["error", { "when": "never", "children": true }],
```
After playing around with this for a bit, I think this setting is more consistent with existing CC practice. See below for an example.

File: .eslintrc.js [hunk 3]
```diff
@@ -1,93 +1,120 @@
...
+    "react/jsx-no-useless-fragment": "error",
+    "react/no-access-state-in-setstate": "error",
+    "react/no-danger": "error",
+    "react/no-unsafe": ["off", { checkAliases: true }],
+    "react/no-unused-state": "error",
+    "react/prop-types": "off",
+    "semi": ["error", "always"]
+  },
+  overrides: [
+    { // rules specific to Jest tests
```
Comment (kswenson) [inline .eslintrc.js:81]: 👍 Nice! Handling the jest and cypress files as overrides is a clear improvement.

File: package.json [hunk 4]
```diff
@@ -37,9 +37,9 @@
     "start": "webpack serve",
     "build": "npm-run-all lint:build build:webpack",
     "build:webpack": "webpack --mode production",
-    "lint": "eslint \"./src/**/*.{js,jsx,ts,tsx}\"",
-    "lint:build": "eslint -c \".eslintrc.build.js\" \"./src/**/*.{js,jsx,ts,tsx}\"",
-    "lint:fix": "eslint --fix \"./src/**/*.{js,jsx,ts,tsx}\"",
+    "lint": "eslint \"./src/**/*.{js,jsx,ts,tsx}\" \"./cypress/**/*.{js,jsx,ts,tsx}\"",
+    "lint:build": "eslint -c \".eslintrc.build.js\" \"./src/**/*.{js,jsx,ts,tsx}\" \"./cypress/**/*.{js,jsx,ts,tsx}\"",
+    "lint:fix": "eslint --fix \"./src/**/*.{js,jsx,ts,tsx}\" \"./cypress/**/*.{js,jsx,ts,tsx}\"",
```
Comment (kswenson) [inline package.json:42]: Nice to get the cypress files handled as well.

File: package.json [hunk 5]
```diff
@@ -66,57 +66,58 @@
...
-    "@cypress/code-coverage": "^3.8.7",
-    "@cypress/webpack-preprocessor": "^5.5.0",
+    "@cypress/code-coverage": "^3.9.2",
+    "@cypress/webpack-preprocessor": "^5.6.0",
     "@istanbuljs/nyc-config-typescript": "^1.0.1",
     "@svgr/webpack": "^5.5.0",
-    "@testing-library/react-hooks": "^3.7.0",
+    "@testing-library/react-hooks": "^5.0.3",
     "@types/enzyme": "^3.10.8",
     "@types/enzyme-adapter-react-16": "^1.0.6",
```
Comment (kswenson) [inline package.json:75]: As you are aware, the `react-16` adapter isn't compatible with React 17, so we should either update the adapter to the unofficial `react-17` adapter (as you've done recently elsewhere) or stick with React 16 until other libraries catch up (as I've done in projects like CLUE).

File: src/components/text.tsx [hunk 6]
```diff
@@ -4,6 +4,6 @@ interface IProps {
   text: string;
 }
 
-export const Text: React.FC<IProps> = ({text}) => (
-  <div>{text}</div>
+export const Text: React.FC<IProps> = ({ text }) => (
+  <div>{ text }</div>
```
Comment (kswenson) [inline src/components/text.tsx:8]: I believe adding spaces here is inconsistent with general CC practice. Changing the configuration from
```
"react/jsx-curly-spacing": ["error", { "when": "never", "children": { "when": "always" } }]
```
to
```
"react/jsx-curly-spacing": ["error", { "when": "never", "children": true }]
```
should allow what seems to me like the more widespread usage.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Mostly this looks good. The linting of the test files is clearly an improvement. I'm ambivalent about whether it's a good idea to mandate particular style guidelines, but it probably doesn't hurt to put them in the starter project as a suggestion with the understanding that they could be relaxed in individual projects when appropriate. Besides, the style guidelines you've chosen are generally consistent with my sense of CC common practice (as well as my personal preferences 😉). The recommended changes, each of which is also described inline, would be:
- synchronize the versions of React and the enzyme adapter
- alphabetize the new rules in the `.eslintrc.js`
- tweak the configuration of the `react/jsx-curly-spacing` rule

**Comment by pjanik** [comment]
> 👍 Mostly this looks good. The linting of the test files is clearly an improvement. I'm ambivalent about whether it's a good idea to mandate particular style guidelines, but it probably doesn't hurt to put them in the starter project as a suggestion with the understanding that they could be relaxed in individual projects when appropriate. Besides, the style guidelines you've chosen are generally consistent with my sense of CC common practice (as well as my personal preferences 😉). 

Thanks. I don't really mind which style we use, but I think it helps not to have to make a micro choice each time. :wink: And it's nice to be able to fix all the inconsistencies automatically with lint --fix. 

> The recommended changes, each of which is also described inline, would be:
> 
> * synchronize the versions of React and the enzyme adapter

Done. I've moved back to React 16, as it might be more obvious for another developer to simply bump React version to 17 rather than understand the thing about unofficial enzyme adapter.

> * alphabetize the new rules in the `.eslintrc.js`

Done.

> * tweak the configuration of the `react/jsx-curly-spacing` rule

I'm not sure which style is more common in fact. I've based my choice simply on the style we have in ActivityPlayer and question-interatives (two most recent JS/TS projects we have / I know). I can skip question-interactives that was mostly setup by me. ;) But it seems that AP also follows the style that is set in this .eslint config, for example:
https://github.com/concord-consortium/activity-player/blob/master/src/components/app.tsx
Again, I'm fine with either, just tried to adjust to what we have now. Also, spaces before `{` and `}` seem to be a bit more consistent with `"object-curly-spacing": ["error", "always"],`. But if you have a stronger opinion about that, I have no problem changing this rule.

**Comment by kswenson** [comment]
Sorry to take so long to get back to you. I've been ruminating on these issues in the background to see how I feel about them.

>>synchronize the versions of React and the enzyme adapter
>
>Done. I've moved back to React 16, as it might be more obvious for another developer to simply bump React version to 17 rather than understand the thing about unofficial enzyme adapter.

This is fine, although thinking about it further an even better solution moving forward would be to eliminate enzyme in favor of `@testing-library/react` and the `starter-projects` would be a good place to start -- but that can be the topic for another PR.

>>tweak the configuration of the react/jsx-curly-spacing rule
>
>I'm not sure which style is more common in fact.

And here we get to the nub of the matter. I doubt there would be a consensus among CC engineers about the right setting for this rule. Furthermore, I'm certain there are engineers (like me) for whom there is no perfect setting for this rule. Additional white space can enhance readability in some contexts and detract from it in others. Personally, I prefer to make that choice in context rather than having it "mandated" for me.

In the end I find I'm not completely comfortable with mandating/suggesting these style rules company-wide. On the other hand I completely respect your (and presumably others) desire to not have to think about things like this. My suggestion would be to move some of these more opinionated stylistic rules into a separate configuration with a couple of additional scripts -- perhaps `lint:style` and `lint:style:fix` -- which can be applied when desired but would not be enabled by default. Perhaps over time a consensus would emerge.

**Comment by mklewandowski** [comment]
> Sorry to take so long to get back to you. I've been ruminating on these issues in the background to see how I feel about them.
> 
> > > synchronize the versions of React and the enzyme adapter
> > 
> > 
> > Done. I've moved back to React 16, as it might be more obvious for another developer to simply bump React version to 17 rather than understand the thing about unofficial enzyme adapter.
> 
> This is fine, although thinking about it further an even better solution moving forward would be to eliminate enzyme in favor of `@testing-library/react` and the `starter-projects` would be a good place to start -- but that can be the topic for another PR.
> 
> > > tweak the configuration of the react/jsx-curly-spacing rule
> > 
> > 
> > I'm not sure which style is more common in fact.
> 
> And here we get to the nub of the matter. I doubt there would be a consensus among CC engineers about the right setting for this rule. Furthermore, I'm certain there are engineers (like me) for whom there is no perfect setting for this rule. Additional white space can enhance readability in some contexts and detract from it in others. Personally, I prefer to make that choice in context rather than having it "mandated" for me.
> 
> In the end I find I'm not completely comfortable with mandating/suggesting these style rules company-wide. On the other hand I completely respect your (and presumably others) desire to not have to think about things like this. My suggestion would be to move some of these more opinionated stylistic rules into a separate configuration with a couple of additional scripts -- perhaps `lint:style` and `lint:style:fix` -- which can be applied when desired but would not be enabled by default. Perhaps over time a consensus would emerge.

Chiming in with my 2 cents:
- I'm all in favor of eventually trying to remove Enzyme and use React Testing Library in the Starter Projects.  I also would like to add test coverage to the translate module so the test coverage starting point when using a tool like CodeCov is 100%.  Granted that is out of scope here.
- I'm also not comfortable with the spacing rule.  I think it depends on context.  I think we should be cautious about adding too many initial linting rules that constrain or even annoy devs.

**Comment by pjanik** [comment]
> This is fine, although thinking about it further an even better solution moving forward would be to eliminate enzyme in favor of @testing-library/react and the starter-projects would be a good place to start -- but that can be the topic for another PR.

Right, I think it could be another PR. Personally, I don't have much experience with `@testing-library/react` and enjoy using Enzyme, but if that's a general trend, I'm totally fine with it. 👍 

> In the end I find I'm not completely comfortable with mandating/suggesting these style rules company-wide. On the other hand I completely respect your (and presumably others) desire to not have to think about things like this. My suggestion would be to move some of these more opinionated stylistic rules into a separate configuration with a couple of additional scripts -- perhaps lint:style and lint:style:fix -- which can be applied when desired but would not be enabled by default. Perhaps over time a consensus would emerge.

Fair. I just put more weight into consistency and lack of micro decision-making (or pointing out whitespaces in PRs), but I totally see your point. 😉 

> My suggestion would be to move some of these more opinionated stylistic rules into a separate configuration with a couple of additional scripts -- perhaps lint:style and lint:style:fix -- which can be applied when desired but would not be enabled by default. Perhaps over time a consensus would emerge.

I like this idea. I've just done that in the last commit. Note that basic .eslint.rc also has some style rules (semicolons, " vs ', etc.), but they've been there forever and didn't seem to cause these issues, so I just left them there. I've only moved there new, opinionated rules that I've added in this PR.

**Comment by pjanik** [comment]
@kswenson, @mklewandowski, any chances you think this could be merged in now? I'm just starting a new project (sound waves) that @mklewandowski might work on. 😉 I think I've removed some of the controversial rules and styling is moved into separate :style NPM scripts.

**Review by kswenson — APPROVED** [review]
👍  LGTM -- I thought we had approved this a while ago. Thx for bringing it up again.

**Pull request merged**